File: automation-test-playwright/pages/customer/CustomerTransactionHistoryPage.ts
import { type Locator } from "@playwright/test";
import { CrudListPage } from "../core/CrudListPage";

export class CustomerTransactionHistoryPage extends CrudListPage {
  protected readonly path = "/customer/transaction/history";

  private readonly monthSelect = this.page.locator('[name="month"]');
  private readonly yearSelect = this.page.locator('[name="year"]');
  private readonly transactionFilterForm = this.page.locator("#filterForm");
  private readonly transactionTableBody = this.page.locator("#transactionTableBody");
  private readonly resultBanner = this.page.locator("#totalTransactionSearch");
  private readonly pagination = this.page.locator("#transactionPagination");

  async waitForLoaded(): Promise<void> {
    await this.page.waitForURL(/\/customer\/transaction\/history/);
    await this.waitForVisible(this.transactionFilterForm);
    await this.waitForVisible(this.transactionTableBody);
  }

  async waitForSummaryVisible(): Promise<void> {
    await this.waitForVisible(this.page.locator(".stats-bar"));
    await this.waitForLocatorCount(this.page.locator(".stats-bar .stat-item"), 2);
  }

  async waitForSummaryValues(totalTransactions: number, totalAmountText: string): Promise<void> {
    const stats = this.page.locator(".stats-bar .stat-item");
    await this.waitForLocatorText(stats.nth(0), String(totalTransactions));
    await this.waitForLocatorText(stats.nth(1), totalAmountText);
  }

  /**
   * Applies the month filter.
   */
  async filterByMonth(month: number | string): Promise<void> {
    await this.monthSelect.selectOption(String(month));
  }

  /**
   * Applies the year filter.
   */
  async filterByYear(year: number | string): Promise<void> {
    await this.yearSelect.selectOption(String(year));
  }

  /**
   * Submits the active filters.
   */
  async submitFilters(): Promise<void> {
    await this.firstVisible(this.transactionFilterForm.locator(".btn-filter, button[type='submit']")).click();
  }

  /**
   * Resets the active filters.
   */
  async resetFilters(): Promise<void> {
    await this.firstVisible(this.transactionFilterForm.locator(".btn-reset, button[type='button']")).click();
  }

  async waitForResultCountBanner(total: number): Promise<void> {
    await this.waitForVisible(this.resultBanner);
    await this.waitForLocatorText(this.resultBanner, String(total));
  }

  rowByBuildingName(buildingName: string): Locator {
    return this.firstVisible(this.transactionTableBody.locator("tr").filter({ hasText: buildingName }));
  }

  async openTransactionDetail(buildingName: string): Promise<void> {
    await this.rowByBuildingName(buildingName).click();
  }

  async waitForDetailModalContains(text: string | RegExp): Promise<void> {
    const modal = this.page.locator(".modal.show");
    await this.waitForVisible(modal);
    const modalText = await this.locatorLooseText(modal);
    if (typeof text === "string") {
      if (!modalText.includes(this.normalizeLooseText(text))) {
        throw new Error(`Transaction detail modal did not contain expected text: ${text}`);
      }
      return;
    }

    const normalizedPattern = new RegExp(this.normalizeLooseText(text.source), text.flags.replace("g", ""));
    if (!normalizedPattern.test(modalText)) {
      throw new Error(`Transaction detail modal did not match expected text: ${String(text)}`);
    }
  }

  /**
   * Closes the detail modal UI element.
   */
  async closeDetailModal(): Promise<void> {
    const modal = this.page.locator(".modal.show");
    await this.firstVisible(modal.locator(".modal-header .btn-close, .modal-footer button")).click();
    await this.waitForHidden(modal);
  }

  async waitForEmptyState(): Promise<void> {
    await this.waitForCondition(
      async () => /khong co giao dich nao/i.test(await this.locatorLooseText(this.transactionTableBody)),
      "Transaction empty state was not displayed."
    );
  }

  async waitForPaginationHidden(): Promise<void> {
    await this.waitForEmpty(this.pagination);
  }
}

File: automation-test-selenium/src/main/java/org/suntower/pages/customer/CustomerTransactionHistoryPage.java
package org.suntower.pages.customer;

import org.openqa.selenium.By;
import org.openqa.selenium.JavascriptExecutor;
import org.openqa.selenium.WebDriver;
import org.openqa.selenium.WebElement;
import org.openqa.selenium.support.ui.ExpectedConditions;
import org.suntower.pages.components.TableComponent;
import org.suntower.pages.core.CrudListPage;

public class CustomerTransactionHistoryPage extends CrudListPage {
  private final By filterForm = css("#filterForm");
  private final By tableBody = css("#transactionTableBody");
  private final By resultBanner = css("#totalTransactionSearch");
  private final By pagination = css("#transactionPagination");
  private final By detailModal = css(".modal.show");
  private final TableComponent table;

  public CustomerTransactionHistoryPage(WebDriver driver) {
    super(driver);
    this.path = "/customer/transaction/history";
    this.table = new TableComponent(driver, "#transactionTableBody");
  }

  public void waitForLoaded() {
    waitForUrlContains("/customer/transaction/history");
    waitForVisible(filterForm);
    waitForVisible(tableBody);
  }

  public void waitForSummaryVisible() {
    waitForVisible(css(".stats-bar"));
    waitForCondition(() -> driver.findElements(css(".stats-bar .stat-item")).size() >= 2, "Transaction stats did not load.");
  }

  public void waitForResultCountBanner(int total) {
    waitForText(resultBanner, String.valueOf(total));
  }

  public void filterByMonth(int month) {
    setFormControlValue("month", String.valueOf(month));
  }

  public void filterByYear(int year) {
    setFormControlValue("year", String.valueOf(year));
  }

  public void submitFilters() {
    wait.until(ExpectedConditions.presenceOfElementLocated(filterForm));
    ((JavascriptExecutor) driver).executeScript("document.querySelector('#filterForm').requestSubmit();");
    table.waitForDataOrEmpty();
  }

  public void resetFilters() {
    ((JavascriptExecutor) driver).executeScript("document.querySelector('#filterForm').reset();");
  }

  public WebElement rowByBuildingName(String buildingName) {
    return table.rowByText(buildingName);
  }

  public void openTransactionDetail(String buildingName) {
    click(rowByBuildingName(buildingName));
  }

  public void waitForDetailModalContains(String text) {
    waitForVisible(detailModal);
    waitForText(detailModal, text);
  }

  public void closeDetailModal() {
    click(firstVisible(css(".modal.show .btn-close, .modal.show .modal-footer button")));
    waitForHidden(detailModal);
  }

  public void waitForEmptyState() {
    waitForTextMatches(tableBody, "khong co giao dich|không có giao dịch");
  }

  public void waitForPaginationHidden() {
    waitForCondition(() -> driver.findElement(pagination).getText().isBlank(), "Transaction pagination was not hidden.");
  }

  private void setFormControlValue(String fieldName, String value) {
    wait.until(ExpectedConditions.presenceOfElementLocated(inputByName(fieldName)));
    ((JavascriptExecutor) driver)
        .executeScript(
            "const el = document.querySelector('[name=\"' + arguments[0] + '\"]');"
                + "if (el.tagName === 'SELECT' && !Array.from(el.options).some(option => option.value === arguments[1])) { el.add(new Option(arguments[1], arguments[1])); }"
                + "el.value = arguments[1];"
                + "el.dispatchEvent(new Event('input', { bubbles: true }));"
                + "el.dispatchEvent(new Event('change', { bubbles: true }));",
            fieldName,
            value);
  }
}
